### ISSUE TITLE
Node not in graph

### ISSUE THREAD

**Issue by PARODBE** [pgmpy/pgmpy#1462]
### Subject of the issue
Hi, I am using a tree augmented naive bayes to solve a problem. And when I go to do inference it tells me that there is not a variable in the node, however when I use the get_cpd function, if I see it. Can you help me to interpret this error, please? Is there any option to force the algorithms to have in does a variable count that has been removed by the model?

### Your environment
* pgmpy version=0.1.15
* Python version=3.8.10
* Operating System=windows 10

### Steps to reproduce
from pgmpy.estimators import TreeSearch
from pgmpy.models import BayesianModel
import networkx as nx

est = TreeSearch(df)
dag = est.estimate(estimator_type="tan",class_node='Target')
bn = BayesianModel(dag.edges())
bn.fit(df)

bn_infer = VariableElimination(bn)
a = bn_infer.query(variables=['Target'], evidence={'var_1':'Si'})
print(a)

### Expected behaviour
Tell us what should happen

### Actual behaviour
Tell us what happens instead

**Comment by ankurankan** [comment]
@PARODBE Would it be possible to share your dataset or the learned tree structure so that I can reproduce the error?

**Comment by PARODBE** [comment]
No...sorry is a private dataset...I understand the problem in order to resolve a problem without dataset....

**Comment by ankurankan** [comment]
Okay, my best guess is that this is probably a bug in network pruning: https://github.com/pgmpy/pgmpy/blob/de1265efcf35a5855ff5cac5a6abd709501b0ee5/pgmpy/inference/base.py#L117. Variable Elimination tries to reduce the size of the network for optimization by removing nodes that don't have any effect on the query. I am guessing it is somehow removing one/more nodes that are required for the query. 

If you are unable to get Variable Elimination to work, you can also do an approximate query by using the `BayesianNetwork.simulate` method.

**Comment by PARODBE** [comment]
But with this function can you put the variables and the evidences? I am reading and only the evidence can be? And why you have to select the number of samples??? Could you show any example of this please using a tan method? And aren't there any more option instead of variable elimination in order to do a query? By the way, I am obtaining a different results than in bnlearn library using also a tan method...

**Comment by ankurankan** [comment]
Here's an example using some data generated from the alarm model. The simulate method doesn't do queries by itself but generates data from the model under various conditions, which can then be used to compute the probabilities of different variables:
```python
# Generate some data to learn a tree and do inference on.
from pgmpy.utils import get_example_model
model = get_example_model('alarm')
data = model.simulate(int(1e5))

# Learn a tree structure from data
from pgmpy.estimators import TreeSearch
est = TreeSearch(data)
tree = est.estimate()

# Learn model parameters
from pgmpy.models import BayesianNetwork
bn = BayesianNetwork(tree.edges())
bn.fit(data)

# Do inference using simulations: P(HISTORY | PAP=NORMAL)
query_data = bn.simulate(int(1e4), evidence={"PAP": "NORMAL"})
query_prob = query_data["HISTORY"].value_counts() / query_data.shape[0]
print(query_prob)
```

**Comment by PARODBE** [comment]
In version 0.1.15 of pgmpy I don't have function simulate...is it possible? Should I have another version?

**Comment by ankurankan** [comment]
@PARODBE Yes, it has not been released yet. You will have to install the dev branch from github: https://pgmpy.org/started/install.html

**Comment by PARODBE** [comment]
In what branch can i find it? And it's possible activate a branch since jupyter notebook? I have never done it...

**Comment by PARODBE** [comment]
In what branch can i find it? And it's possible activate a branch since jupyter notebook? I have never done it...

**Comment by ankurankan** [comment]
You will have to reinstall pgmpy. You can use pip to install the dev branch from github. Something like this:
```bash
pip uninstall pgmpy
pip install git+https://github.com/pgmpy/pgmpy.git@dev
```

**Comment by PARODBE** [comment]
Thanks, and actually, what wiil be the main difference between using variableelimination? That don't remove any variable?

**Comment by ankurankan** [comment]
@PARODBE Both are completely different ways to do inference. Variable Elimination is one specific algorithm that systematically sums out variables in the model to get to the final results whereas the simulation method generates data from the model which we then use to compute the probabilities. The main practical difference is that Variable Elimination is an exact inference algorithm whereas the simulation method will give approximate results (which asymptotically approaches the exact solution as the number of samples is increased)
